### PR TITLE
support AFL++ fuzzing

### PR DESCRIPTION
**Pull request by jlevon**

### CODE REVIEW

File: samples/server.c [hunk 1]
```diff
@@ -461,7 +462,11 @@ int main(int argc, char *argv[])
         err(EXIT_FAILURE, "failed to register signal handler");
     }
 
-    vfu_ctx = vfu_create_ctx(VFU_TRANS_SOCK, argv[optind], 0, &server_data,
+    if (strcmp(argv[optind], "pipe") == 0) {
```
Comment (tmakatos) [inline samples/server.c:465]: Why do we support the pipe transport on the server but not on the client?

**Review by tmakatos — COMMENTED** [review]
Why can't the existing, UNIX domain socket transport be used for fuzzing?

I see some duplication between tran_sock and tran_pipe, could we avoid that by moving the tran_pipe stuff into tran_sock?

Or maybe implement a simple Python/bash script that relays messages receives over the UNIX domain socket to the pipe and vice versa?

File: samples/server.c (hunk 1 above)
Comment (jlevon) [reply samples/server.c:465]: afl is the client

File: docs/testing.md [hunk 2]
```diff
@@ -15,3 +15,38 @@ The master branch is run through [Coverity](scan.coverity.com) when a new PR
...
+docker run -ti -v /path/to/libvfio-user:/src aflplusplus/aflplusplus
+```
+
+Set up and build:
+
+```
+apt update
+apt-get -y install libjson-c-dev libcmocka-dev clang valgrind python3-pytest debianutils flake8 libssl-dev cmake
+
+cd /src
```
[the remark this answers is not in the document]
Comment (jlevon) [reply docs/testing.md:36]: did you pass the right -v argument to docker?

**Comment by jlevon** [comment]
> Why can't the existing UNIX domain socket transport be used for fuzzing?
> 

AFL expects to be able to start a process and write to it for fuzzing, so going through unix domain socket doesn't work (easily).

> I see some duplication between tran_sock and tran_pipe, could we avoid that by moving the tran_pipe stuff into tran_sock?

Yeah, at least negotiate has a ton of duplication, let me take a look at fixing this.

>

**Comment by jlevon** [comment]
Thanos: I refactored to share the negotiate code. I think trying to go any further would actually complicate things more - I don't want to add too much more to tran API.

**Comment by swapnili** [comment]
I do not like the fact that this code around VFU_TRANS_PIPE will be part of regular builds although used only for testing. 
Can we somehow extract it out and build only in case for AFL testing?

**Comment by swapnili** [comment]
> I don't want this to regress (not build) as we make changes though. I could add a build option, and only set it for "make pre-push" (or if explicitly given) ? Would that be better?

Yes that would be better to build only in case if explicitly asked or pre-push. Thanks

**Comment by jlevon** [comment]
@swapnili @tmakatos ok now?

File: lib/tran_pipe.c [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+    tran_pipe_t *tp = NULL;
+    int ret = 0;
+
+    assert(vfu_ctx != NULL);
+
+    tp = calloc(1, sizeof(tran_pipe_t));
+
+    if (tp == NULL) {
+        ret = errno;
+        goto out;
```
Comment (swapnili) [inline lib/tran_pipe.c:254]: This goto is not required you can simply free() and return from here?

**Review by swapnili — APPROVED** [review]
Looks good with one minor suggestion. Thanks

File: lib/tran_pipe.c [hunk 4]
```diff
@@ -250,14 +250,12 @@ tran_pipe_init(vfu_ctx_t *vfu_ctx)
     tp = calloc(1, sizeof(tran_pipe_t));
 
     if (tp == NULL) {
-        ret = errno;
-        goto out;
+        return -1;
```
Comment (swapnili) [inline lib/tran_pipe.c:253]: return ERROR_INT(errno)

File: lib/tran_sock.c [hunk 5]
```diff
@@ -375,8 +375,7 @@ tran_sock_init(vfu_ctx_t *vfu_ctx)
     ts = calloc(1, sizeof(tran_sock_t));
 
     if (ts == NULL) {
-        ret = errno;
-        goto out;
+        return -1;
```
Comment (swapnili) [inline lib/tran_sock.c:378]: ditto

File: lib/tran_pipe.c (hunk 4 above)
Comment (jlevon) [reply lib/tran_pipe.c:253]: why? What's here is the exact same.

Comment (swapnili) [reply lib/tran_pipe.c:253]: Just to be consistent with other returns. But I am fine with this also.

**Pull request merged**